OpenRA.Mods.Common/Widgets/ModelWidget.cs

```diff
@@ -0,0 +1,222 @@
+			var cameraAngle = GetCameraAngle();
+
+			if (voxel == null || palette == null)
+				return;
+
+			if (voxel != cachedVoxel)
+				cachedVoxel = voxel;
+
+			if (palette != cachedPalette)
+			{
+				if (string.IsNullOrEmpty(palette) && string.IsNullOrEmpty(playerPalette))
+					return;
+
+				string paletteName = string.IsNullOrEmpty(palette) ? playerPalette : palette;
+				paletteReference = WorldRenderer.Palette(paletteName);
+				cachedPalette = paletteName;
+			}
+
+			if (playerPalette != cachedPlayerPalette)
+			{
+				paletteReferencePlayer = WorldRenderer.Palette(playerPalette);
+				cachedPlayerPalette = playerPalette;
+			}
+
+			if (normalsPalette != cachedNormalsPalette)
+			{
+				paletteReferenceNormals = WorldRenderer.Palette(normalsPalette);
+				cachedNormalsPalette = normalsPalette;
+			}
+
+			if (shadowPalette != cachedShadowPalette)
+			{
+				paletteReferenceShadow = WorldRenderer.Palette(shadowPalette);
+				cachedShadowPalette = shadowPalette;
+			}
+
+			if (scale != cachedScale)
+				cachedScale = scale;
+
+			if (rotation != cachedRotation)
+				cachedRotation = rotation;
+
+			if (lightPitch != cachedLightPitch)
+				cachedLightPitch = lightPitch;
+
+			if (lightYaw != cachedLightYaw)
+				cachedLightYaw = lightYaw;
+
+			if (cachedLightAmbientColor[0] != lightAmbientColor[0] || cachedLightAmbientColor[1] != lightAmbientColor[1] || cachedLightAmbientColor[2] != lightAmbientColor[2])
+				cachedLightAmbientColor = lightAmbientColor;
+
+			if (cachedLightDiffuseColor[0] != lightDiffuseColor[0] || cachedLightDiffuseColor[1] != lightDiffuseColor[1] || cachedLightDiffuseColor[2] != lightDiffuseColor[2])
+				cachedLightDiffuseColor = lightDiffuseColor;
+
+			if (cameraAngle != cachedCameraAngle)
+				cachedCameraAngle = cameraAngle;
+
+			if (cachedVoxel == null)
+				return;
+
+			var size = new float2(cachedVoxel.Size[0] * cachedScale, cachedVoxel.Size[1] * cachedScale);
+			ModelAnimation animation = new ModelAnimation(
+				cachedVoxel,
+				() => WVec.Zero,
+				() => rotation,
+				() => false,
+				() => 0,
+				true);
+
+			var animations = new ModelAnimation[] { animation };
+
+			ModelPreview preview = new ModelPreview(
+				new ModelAnimation[] { animation }, WVec.Zero, 0,
+				cachedScale,
+				new WAngle(cachedLightPitch),
+				new WAngle(cachedLightYaw),
+				cachedLightAmbientColor,
+				cachedLightDiffuseColor,
+				cachedCameraAngle,
+				paletteReference,
+				paletteReferenceNormals,
+				paletteReferenceShadow);
+
+			var screenBounds = animation.ScreenBounds(WPos.Zero, WorldRenderer, scale);
+			IdealPreviewSize = new int2(screenBounds.Width, screenBounds.Height);
+			PreviewOffset = -new int2(screenBounds.Left, screenBounds.Top) - IdealPreviewSize / 2;
+			var origin = RenderOrigin + new int2(RenderBounds.Size.Width / 2, RenderBounds.Size.Height / 2);
+
+			var modelRenderable = new UIModelRenderable(
+				animations, WPos.Zero, origin - PreviewOffset, 0, WRot.FromYaw(cameraAngle), scale,
```

Comment:
Is it intended that the cameraAngle only affects the yaw angle? If you want anything other than a straight top-down perspective you would need to adjust the roll or pitch angle.

I think it would be the most intuitive  to use the same camera perspective as when the voxel is placed on the map.

Reply:
Optimized the default camera angle.